File: leetcode/00946_validate-stack-sequences/946-validate-stack-sequences.py
class Solution:
    # time O(N), space O(N)
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        idx = 0
        stack = []
        for e in pushed:
            if e == popped[idx]:
                idx += 1
            else:
                stack.append(e)

            while stack and stack[-1] == popped[idx]:
                idx += 1
                stack.pop()

        return len(stack) == 0 and idx == len(popped)

File: leetcode/00946_validate-stack-sequences/946-validate-stack-sequences.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        vector<int> stack;
        int idx = 0;
        int length = pushed.size();
        for (int i = 0; i < length; i++) {
            if (pushed[i] == popped[idx]) {
                idx += 1;
            } else {
                stack.push_back(pushed[i]);
            }
            
            while (idx < length && stack.size() && stack.back() == popped[idx]) {
                stack.pop_back();
                idx += 1;
            }
        }
        return idx == length;
    }
};